source-test/File.ts

```diff
@@ -1,661 +1,1617 @@
+      ];
+
+      const file = new Vinyl(options);
+      file.path = history[1];
+      file.path = history[2];
+      const file2 = file.clone();
+
+      expect(file2.history).toEqual(history);
+      expect(file2.history).toNotBe(file.history);
+      expect(file2.path).toEqual(history[2]);
+      done();
     });
 
-    it('should do nothing with null', done => {
+    it('supports deep & shallow copy of all attributes', function(done) {
       const options = {
         cwd: '/',
         base: '/test/',
         path: '/test/test.coffee',
-        contents: fakeStream
+        contents: <Buffer> null,
+        custom: { meta: {} },
       };
-      const file = new File(options);
-      const stream = new Stream.PassThrough();
-      stream.on('data', () => {
-        throw new Error('should not write');
-      });
-      stream.on('end', () => {
-        throw new Error('should not end');
+
+      const file: Vinyl & {custom: { meta: {} }} = <any> new Vinyl(options);
+
+      const file2: Vinyl & {custom: { meta: {} }} = <any> file.clone();
+      expect(file2.custom).toEqual(file.custom);
+      expect(file2.custom).toNotBe(file.custom);
+      expect(file2.custom.meta).toEqual(file.custom.meta);
+      expect(file2.custom.meta).toNotBe(file.custom.meta);
+
+      const file3: Vinyl & {custom: { meta: {} }} = <any> file.clone(true);
+      expect(file3.custom).toEqual(file.custom);
+      expect(file3.custom).toNotBe(file.custom);
+      expect(file3.custom.meta).toEqual(file.custom.meta);
+      expect(file3.custom.meta).toNotBe(file.custom.meta);
+
+      const file4: Vinyl & {custom: { meta: {} }} = <any> file.clone({ deep: true });
+      expect(file4.custom).toEqual(file.custom);
+      expect(file4.custom).toNotBe(file.custom);
+      expect(file4.custom.meta).toEqual(file.custom.meta);
+      expect(file4.custom.meta).toNotBe(file.custom.meta);
+
+      const file5: Vinyl & {custom: { meta: {} }} = <any> file.clone(false);
+      expect(file5.custom).toEqual(file.custom);
+      expect(file5.custom).toBe(file.custom);
+      expect(file5.custom.meta).toEqual(file.custom.meta);
+      expect(file5.custom.meta).toBe(file.custom.meta);
+
+      const file6: Vinyl & {custom: { meta: {} }} = <any> file.clone({ deep: false });
+      expect(file6.custom).toEqual(file.custom);
+      expect(file6.custom).toBe(file.custom);
+      expect(file6.custom.meta).toEqual(file.custom.meta);
+      expect(file6.custom.meta).toBe(file.custom.meta);
+
+      done();
+    });
+
+    it('supports inheritance', function(done) {
+      function ExtendedFile() {
+        Vinyl.apply(this, arguments);
+      }
+      ExtendedFile.prototype = Object.create(Vinyl.prototype);
+      ExtendedFile.prototype.constructor = ExtendedFile;
+      // Just copy static stuff since Object.setPrototypeOf is node >=0.12
+      Object.keys(Vinyl).forEach((key) => {
+        (<any> ExtendedFile)[key] = (<{[key: string]: any}> Vinyl)[key];
       });
```

Comment:
Since this is TypeScript that gets compiled, you can test this with `class ExtendedFile extends Vinyl {}`